File: ts/src/loader.ts
/** CSV loading for Phase 1, browser-friendly: parses CSV text (no filesystem
 * access here). Statement CSV is pre-parsed (hand-typed from a PDF). */
import type { LedgerLine, StatementLine } from "./models";
import { parseAmount } from "./money";
import { normaliseRef } from "./normalise";

/** Minimal CSV parser: quoted fields with "" escapes, \r\n or \n line ends. */
export function parseCsv(text: string): Record<string, string>[] {
  const rows: string[][] = [];
  let row: string[] = [];
  let field = "";
  let inQuotes = false;
  for (let i = 0; i < text.length; i++) {
    const c = text[i];
    if (inQuotes) {
      if (c === '"') {
        if (text[i + 1] === '"') {
          field += '"';
          i++;
        } else {
          inQuotes = false;
        }
      } else {
        field += c;
      }
    } else if (c === '"') {
      inQuotes = true;
    } else if (c === ",") {
      row.push(field);
      field = "";
    } else if (c === "\n" || c === "\r") {
      if (c === "\r" && text[i + 1] === "\n") i++;
      row.push(field);
      field = "";
      rows.push(row);
      row = [];
    } else {
      field += c;
    }
  }
  if (field !== "" || row.length > 0) {
    row.push(field);
    rows.push(row);
  }
  const header = rows.shift() ?? [];
  return rows
    .filter((r) => r.length > 1 || (r.length === 1 && r[0] !== ""))
    .map((r) => Object.fromEntries(header.map((h, j) => [h, r[j] ?? ""])));
}

function parseDate(raw: string): string {
  const s = raw.trim();
  if (!/^\d{4}-\d{2}-\d{2}$/.test(s)) throw new Error(`bad date: ${raw}`);
  return s;
}

export function loadStatementCsv(text: string): StatementLine[] {
  return parseCsv(text).map((row, i) => {
    const rawRef = (row["ref"] ?? "").trim();
    return {
      id: `S${i + 1}`,
      raw_ref: rawRef,
      normalised_ref: normaliseRef(rawRef),
      doc_date: parseDate(row["date"] ?? ""),
      doc_type: (row["type"] ?? "").trim().toUpperCase(),
      amount: parseAmount(row["amount"] ?? ""),
      currency: (row["currency"] ?? "").trim().toUpperCase() || "USD",
      po_number: (row["po"] ?? "").trim(),
    };
  });
}

export function loadLedgerCsv(text: string): LedgerLine[] {
  return parseCsv(text).map((row, i) => {
    const rawRef = (row["ref"] ?? "").trim();
    return {
      id: `L${i + 1}`,
      supplier: (row["supplier"] ?? "").trim(),
      raw_ref: rawRef,
      normalised_ref: normaliseRef(rawRef),
      doc_date: parseDate(row["date"] ?? ""),
      doc_type: (row["type"] ?? "").trim().toUpperCase(),
      original_amount: parseAmount(row["original"] ?? ""),
      open_amount: parseAmount(row["open"] ?? ""),
      currency: (row["currency"] ?? "").trim().toUpperCase() || "USD",
      po_number: (row["po"] ?? "").trim(),
    };
  });
}

File: src/tieout/loader.py
"""CSV loading for Phase 1. Statement CSV is pre-parsed (hand-typed from a PDF)."""
from __future__ import annotations

import csv
from datetime import date
from pathlib import Path

from .models import LedgerLine, StatementLine
from .money import parse_amount
from .normalise import normalise_ref


def _parse_date(raw: str) -> date:
    return date.fromisoformat(raw.strip())


def load_statement_csv(path: str | Path) -> list[StatementLine]:
    lines: list[StatementLine] = []
    with open(path, newline="", encoding="utf-8") as f:
        for i, row in enumerate(csv.DictReader(f)):
            raw_ref = row["ref"].strip()
            lines.append(
                StatementLine(
                    id=f"S{i + 1}",
                    raw_ref=raw_ref,
                    normalised_ref=normalise_ref(raw_ref),
                    doc_date=_parse_date(row["date"]),
                    doc_type=row["type"].strip().upper(),
                    amount=parse_amount(row["amount"]),
                    currency=row.get("currency", "").strip().upper() or "USD",
                    po_number=(row.get("po") or "").strip(),
                )
            )
    return lines


def load_ledger_csv(path: str | Path) -> list[LedgerLine]:
    lines: list[LedgerLine] = []
    with open(path, newline="", encoding="utf-8") as f:
        for i, row in enumerate(csv.DictReader(f)):
            raw_ref = row["ref"].strip()
            lines.append(
                LedgerLine(
                    id=f"L{i + 1}",
                    supplier=row["supplier"].strip(),
                    raw_ref=raw_ref,
                    normalised_ref=normalise_ref(raw_ref),
                    doc_date=_parse_date(row["date"]),
                    doc_type=row["type"].strip().upper(),
                    original_amount=parse_amount(row["original"]),
                    open_amount=parse_amount(row["open"]),
                    currency=row.get("currency", "").strip().upper() or "USD",
                    po_number=(row.get("po") or "").strip(),
                )
            )
    return lines
